src/libcore/cell.rs

```diff
@@ -1101,13 +1101,13 @@ struct BorrowRef<'b> {
 impl<'b> BorrowRef<'b> {
     #[inline]
     fn new(borrow: &'b Cell<BorrowFlag>) -> Option<BorrowRef<'b>> {
-        let b = borrow.get();
-        if is_writing(b) || b == isize::max_value() {
+        let b = borrow.get().wrapping_add(1);
```

Comment:
what about the `isize::max_value` comparision? I doubt that it will ever hit 16k/4.2M/4.2M² borrows, but who knows ^^

Reply:
This still indirectly checks for max_value. This needs a comment though, and I think the `b` check should be explicitly expanded, as the is_reading check doesn't really make sense semantically anymore.

Reply:
In this case the wrapping_add result will be isize::min_value, which is in the writing range, so the is_reading check will fail and None will be returned like now. This diff is not supposed to change behavior, it's just an optimization.

The point of this change is that instead of checking for min_value <= x < 0 || x == max_value before adding 1, we can equivalently check for min_value < x.wrapping_add(1) <= 0 or x.wrapping_add(1) == (max_value.wrapping_add(1)) == min_value after adding, which simplifies to min_value <= x + 1 <= 0 which is what !is_reading(b) expands to.

Reply:
It would still be good to add a comment though

Reply:
I think the is_reading check makes sense semantically: we want the resulting value to be a value that indicates that the RefCell is being read, so if the result fails is_reading we can't succeed; on the other hand, we should also succeed for all values where is_reading succeeds since all values for the borrow flag are valid and otherwise some would be unobtainable.

Hence, we semantically want to succeed exactly when is_reading on the value we would write succeeds.